# PRADA2

Description

Massively parallel sequencing of cDNA reverse transcribed from RNA (RNASeq) provides an accurate estimate of the quantity and composition of mRNAs. To characterize the transcriptome through the analysis of RNA-seq data, we developed PRADA. PRADA focuses on the processing and analysis of gene expression estimates, supervised and unsupervised gene fusion identification, and supervised intragenic deletion identification. The BAM files generated by the pipeline are readily compatible with different tools for mutation calling and to obtain read counts for further downstream analysis.

## Prerequisites
| Items | Values |
| ----------- | ----------- |
| OS requirements | Linux kernel (Ubuntu, Centos) |
| Memory | 30G(Minimum) |
| Software | Conda |
| Language | Python2.7 |



## Installing

1. **Download repository**

```
git clone https://github.com/juechenyang/PRADA2.git
```

2. **Install conda**

* Download conda bash installer [here](https://docs.conda.io/en/latest/miniconda.html)
* Bash command to install:

<pre>bash<i> your_path_to_Miniconda3-latest-Linux-x86_64.sh</i></pre>


3. **Create prada enviroment**

* locate the file "prada2.yml" inside "env" folder
* create the conda environment using "prada2.yml" 

<pre>conda env create -f <i>your_path_to_prada2.yml</i></pre>


4. **Activate enviroment**

```
conda activate prada
```

## Data prepare

* locate the "download_data.py" script inside the /alignment_inputs
* make sure you have activated prada enviroment in step 4 of installing
* run the following command:

<pre>python <i>your_path_to_download_data.py</i></pre>

After the downloading process finished, there should be two files in /alignment_input:
* GRCh38.d1.vd1.fa (reference genome)
* gencode.v22.annotation.gtf (annotation gtf)


## Executing

* Alignment
<pre>python prada2.py --read1 <i>your_path_to_read1_fastq</i> --read2 <i>your_path_to_read2_fastq</i> --outdir <i>your_path_to_output_dir</i></pre>

* Fusion detection
<pre>python prada2.py --read1 <i>your_path_to_read1_fastq</i> --read2 <i>your_path_to_read2_fastq</i> --outdir <i>your_path_to_output_dir --fusion</i></pre>

* Get rsem gene expression
<pre>python prada2.py --read1 <i>your_path_to_read1_fastq</i> --read2 <i>your_path_to_read2_fastq</i> --outdir <i>your_path_to_output_dir</i> --rsem</pre>

## Authors

* **Juechen Yang** - *Programmer Analyst @ UT Health, San Antonio* 
* **Siyuan Zheng** - *Assistant Professor @ UT Health, San Antonio* - [our team](http://zhenglab.info/)

## License

* **MIT**
